Enhanced the SSA value propagation logic in deabstraction

mhong opened the pull request:
This is to support struct/tuple field values that are not directly linked to tensor op insts via
operands/results.

One example is in propagating `%188` to `%190` below, where `%189` is not a const
struct. This is so that const expr eval will only process the (const) struct
field `%188`, and not the struct `%189`.  `%188` is eventually used (through
`_allocateUninitializedArray()`) by the tfop attr `%181`.

```
  // function_ref _allocateUninitializedArray<A>(_:)
  %179 = function_ref @$Ss27_allocateUninitializedArrayySayxG_BptBwlF : $@convention(thin) <�_0_0> (Builtin.Word) -> (@owned Array<�_0_0>, Builtin.RawPointer) // user: %180
  %180 = apply %179<TensorShape>(%178) : $@convention(thin) <�_0_0> (Builtin.Word) -> (@owned Array<�_0_0>, Builtin.RawPointer) // users: %184, %183, %181
  %181 = tuple_extract %180 : $(Array<TensorShape>, Builtin.RawPointer), 0 // users: %194, %193, %182
  retain_value %181 : $Array<TensorShape>         // id: %182
  %183 = tuple_extract %180 : $(Array<TensorShape>, Builtin.RawPointer), 1 // user: %185
  release_value %180 : $(Array<TensorShape>, Builtin.RawPointer) // id: %184
  %185 = pointer_to_address %183 : $Builtin.RawPointer to [strict] $*TensorShape // user: %192
  %186 = struct $_ContiguousArrayBuffer<Int32> (%149 : $_ContiguousArrayStorageBase) // user: %187
  %187 = struct $Array<Int32> (%186 : $_ContiguousArrayBuffer<Int32>) // user: %188
  %188 = struct $TensorShape (%187 : $Array<Int32>) // user: %189
  %189 = struct $SimpleIterator (%145 : $ResourceHandle, %188 : $TensorShape) // user: %190
  %190 = struct_extract %189 : $SimpleIterator, #SimpleIterator.elementShape // users: %192, %191
  retain_value %190 : $TensorShape                // id: %191
  store %190 to %185 : $*TensorShape              // id: %192
  %193 = builtin "__tfop_IteratorGetNext,$in,output_types,output_shapes"(%137 : $ResourceHandle, %170 : $Array<Int32.Type>, %181 : $Array<TensorShape>) : $Tensor<Float> // user: %204
```

This is done by calling `arrayDecoder.decode()` within `propagateSSAOperand()`.

Also, to make sure that we can propagate values of `$TensorShape` struct type:
1. In `explodeAggregateInst()`, changed `emitLoweredXXXMostDerivedDescendents()`
calls to `emitLoweredXXXDirectChildren()`, so that we do not dig into and
explode structs that are unrelated to tensorflow value types. One such example
is $Array<Int32>, which is a struct field of $TensorShape.

2. Extended SROA to allow skipping the explosion of `alloc_stack` over the
`$TensorShape` type.

Without the above changes, we'll be dealing with exploded low-level structs such as `%186` and `%187` below (the guts of the array struct), and that would fail deabstraction, as we cannot reverse-trace `%188` through these structs back to `%146`.

```
  %144 = struct $SimpleIterator (%137 : $ResourceHandle, %142 : $TensorShape) // users: %146, %145
  %145 = struct_extract %144 : $SimpleIterator, #SimpleIterator.handle // users: %209, %189, %153
  %146 = struct_extract %144 : $SimpleIterator, #SimpleIterator.elementShape // user: %147
  %147 = struct_extract %146 : $TensorShape, #TensorShape.dimensions // user: %148
  %148 = struct_extract %147 : $Array<Int32>, #Array._buffer // user: %149
  %149 = struct_extract %148 : $_ContiguousArrayBuffer<Int32>, #_ContiguousArrayBuffer._storage // users: %206, %186, %150
  ...
  %178 = integer_literal $Builtin.Word, 1         // user: %180
  // function_ref _allocateUninitializedArray<A>(_:)
  %179 = function_ref @$Ss27_allocateUninitializedArrayySayxG_BptBwlF : $@convention(thin) <Ä_0_0> (Builtin.Word) -> (@owned Array<Ä_0_0>, Builtin.RawPointer) // user: %180
  %180 = apply %179<TensorShape>(%178) : $@convention(thin) <Ä_0_0> (Builtin.Word) -> (@owned Array<Ä_0_0>, Builtin.RawPointer) // users: %184, %183, %181
  %181 = tuple_extract %180 : $(Array<TensorShape>, Builtin.RawPointer), 0 // users: %194, %193, %182
  retain_value %181 : $Array<TensorShape>         // id: %182
  %183 = tuple_extract %180 : $(Array<TensorShape>, Builtin.RawPointer), 1 // user: %185
  release_value %180 : $(Array<TensorShape>, Builtin.RawPointer) // id: %184
  %185 = pointer_to_address %183 : $Builtin.RawPointer to [strict] $*TensorShape // user: %192
  %186 = struct $_ContiguousArrayBuffer<Int32> (%149 : $_ContiguousArrayStorageBase) // user: %187
  %187 = struct $Array<Int32> (%186 : $_ContiguousArrayBuffer<Int32>) // user: %188
  %188 = struct $TensorShape (%187 : $Array<Int32>) // user: %189
  %189 = struct $SimpleIterator (%145 : $ResourceHandle, %188 : $TensorShape) // user: %190
  %190 = struct_extract %189 : $SimpleIterator, #SimpleIterator.elementShape // users: %192, %191
  retain_value %190 : $TensorShape                // id: %191
  store %190 to %185 : $*TensorShape              // id: %192
  %193 = builtin "__tfop_IteratorGetNext,$in,output_types,output_shapes"(%137 : $ResourceHandle, %170 : $Array<Int32.Type>, %181 : $Array<TensorShape>) : $Tensor<Float> // user: %204
```

While we may need to further enhance deabstraction code to propagate through
more SIL patterns (especially related to const array handling), the code is
structured in a way that's reasonably easy to extend.


This is a follow-up to https://github.com/apple/swift/pull/18650.
Resolves [SR-8463](https://bugs.swift.org/browse/SR-8463).

Commit:
Enhanced the SSA value propagation logic in deabstraction, to support
struct/tuple field values that are not directly linked to tensor op insts via
operands/results.

One example is in propagating %188 to %190 below, where %189 is not a const
struct. This is so that const expr eval will only process the (const) struct
field %188, and not the struct %189.  %188 is eventually used (through
_allocateUninitializedArray()) by the tfop attr %181.

```
  // function_ref _allocateUninitializedArray<A>(_:)
  %179 = function_ref @$Ss27_allocateUninitializedArrayySayxG_BptBwlF : $@convention(thin) <�_0_0> (Builtin.Word) -> (@owned Array<�_0_0>, Builtin.RawPointer) // user: %180
  %180 = apply %179<TensorShape>(%178) : $@convention(thin) <�_0_0> (Builtin.Word) -> (@owned Array<�_0_0>, Builtin.RawPointer) // users: %184, %183, %181
  %181 = tuple_extract %180 : $(Array<TensorShape>, Builtin.RawPointer), 0 // users: %194, %193, %182
  retain_value %181 : $Array<TensorShape>         // id: %182
  %183 = tuple_extract %180 : $(Array<TensorShape>, Builtin.RawPointer), 1 // user: %185
  release_value %180 : $(Array<TensorShape>, Builtin.RawPointer) // id: %184
  %185 = pointer_to_address %183 : $Builtin.RawPointer to [strict] $*TensorShape // user: %192
  %186 = struct $_ContiguousArrayBuffer<Int32> (%149 : $_ContiguousArrayStorageBase) // user: %187
  %187 = struct $Array<Int32> (%186 : $_ContiguousArrayBuffer<Int32>) // user: %188
  %188 = struct $TensorShape (%187 : $Array<Int32>) // user: %189
  %189 = struct $SimpleIterator (%145 : $ResourceHandle, %188 : $TensorShape) // user: %190
  %190 = struct_extract %189 : $SimpleIterator, #SimpleIterator.elementShape // users: %192, %191
  retain_value %190 : $TensorShape                // id: %191
  store %190 to %185 : $*TensorShape              // id: %192
  %193 = builtin "__tfop_IteratorGetNext,$in,output_types,output_shapes"(%137 : $ResourceHandle, %170 : $Array<Int32.Type>, %181 : $Array<TensorShape>) : $Tensor<Float> // user: %204
```

Another example is in propagating %142 to %146, where the struct %144 is not
const. %146 is eventually used as a tfop attr in 181, through a series of struct
wrappers in %186, %187 and %188.

```
  %144 = struct $SimpleIterator (%137 : $ResourceHandle, %142 : $TensorShape) // users: %146, %145
  %145 = struct_extract %144 : $SimpleIterator, #SimpleIterator.handle // users: %209, %189, %153
  %146 = struct_extract %144 : $SimpleIterator, #SimpleIterator.elementShape // user: %147
  %147 = struct_extract %146 : $TensorShape, #TensorShape.dimensions // user: %148
  %148 = struct_extract %147 : $Array<Int32>, #Array._buffer // user: %149
  %149 = struct_extract %148 : $_ContiguousArrayBuffer<Int32>, #_ContiguousArrayBuffer._storage // users: %206, %186, %150
  ...
  %178 = integer_literal $Builtin.Word, 1         // user: %180
  // function_ref _allocateUninitializedArray<A>(_:)
  %179 = function_ref @$Ss27_allocateUninitializedArrayySayxG_BptBwlF : $@convention(thin) <Ä_0_0> (Builtin.Word) -> (@owned Array<Ä_0_0>, Builtin.RawPointer) // user: %180
  %180 = apply %179<TensorShape>(%178) : $@convention(thin) <Ä_0_0> (Builtin.Word) -> (@owned Array<Ä_0_0>, Builtin.RawPointer) // users: %184, %183, %181
  %181 = tuple_extract %180 : $(Array<TensorShape>, Builtin.RawPointer), 0 // users: %194, %193, %182
  retain_value %181 : $Array<TensorShape>         // id: %182
  %183 = tuple_extract %180 : $(Array<TensorShape>, Builtin.RawPointer), 1 // user: %185
  release_value %180 : $(Array<TensorShape>, Builtin.RawPointer) // id: %184
  %185 = pointer_to_address %183 : $Builtin.RawPointer to [strict] $*TensorShape // user: %192
  %186 = struct $_ContiguousArrayBuffer<Int32> (%149 : $_ContiguousArrayStorageBase) // user: %187
  %187 = struct $Array<Int32> (%186 : $_ContiguousArrayBuffer<Int32>) // user: %188
  %188 = struct $TensorShape (%187 : $Array<Int32>) // user: %189
  %189 = struct $SimpleIterator (%145 : $ResourceHandle, %188 : $TensorShape) // user: %190
  %190 = struct_extract %189 : $SimpleIterator, #SimpleIterator.elementShape // users: %192, %191
  retain_value %190 : $TensorShape                // id: %191
  store %190 to %185 : $*TensorShape              // id: %192
  %193 = builtin "__tfop_IteratorGetNext,$in,output_types,output_shapes"(%137 : $ResourceHandle, %170 : $Array<Int32.Type>, %181 : $Array<TensorShape>) : $Tensor<Float> // user: %204
```

mhong commented:
@swift-ci please test tensorflow

Commit:
Polished code.

mhong commented:
@lattner I played with the array decoder based approach, and find it working for example 1 above, but not example 2. That made me go with the more general approach of trying to propagate ssa values for all operands and insts. PTAL.

mhong commented:
CI passed: https://ci-external.swift.org/view/Pull%20Request/job/swift-PR-TensorFlow-Linux/651/

lattner commented:
The second example looks like a different problem.  The patch as you've built looks like it will rip through lots of SIL code and effectively remove all struct/struct extract operations, even if they have nothing to do with TensorFlow APIs.  This isn't appropriate for a guaranteed optimization to do, because it will blow away debug information and other stuff for general swift code.

I'd prefer that you go with the approach we were talking about yesterday, and we can continue to look at that second case after the first is fixed.

bgogul commented on lib/SILOptimizer/Mandatory/TFDeabstraction.cpp:
```diff
@@ -1363,19 +1363,6 @@ propagateTensorOperand(SILValue v,
       }
-
-      // Otherwise simplify inputs in predecessor blocks.
-      for (auto pi : arg->getParent()->getPredecessorBlocks()) {
```
Is this no longer needed because we look at all the operands now? If that is the case, shouldn't the iteration below ensure all its predecessors are processed before a node is processed?

mhong replied:
Reverted this change.

mhong commented:
@lattner I agree that we should not rewrite the operand of `debug_value`. What if we exclude that inst type, and rewrite _all other_ operands that are produced by `struct_extract`, etc? Intuitively, an inst that consumes an operand produced by `struct_extract` should have the same semantic behavior if we rewire the operand to come from some "earlier producer" inst of that struct field.

The above is a "blacklist" based approach, where we blacklist the `debug_value` inst type.

The concern with just tracing through `_allocateUninitializedArray` is that I'm not sure how many other cases we'll need to write code to trace through (example 2 above is another case, and there might be other cases). This is a "whitelist" based approach, and it's not clear to me whether this whitelist can expand to include a large portion of what SIL can express.

(I'll try following up with you in person if it works with your schedule.)

@bgogul, thanks for the comments and questions! Once we converge on the high level design, I'll address them at the code level.

Commit:
Per discussion with @lattner, adjusted this patch to only do ssa value
propagation on SIL insts related to tfop operands (including attrs).

To make sure that we can propagate values of $TensorShape struct type:
1. In `explodeAggregateInst()`, changed `emitLoweredXXXMostDerivedDescendents()`
calls to `emitLoweredXXXDirectChildren()`, so that we do not dig into and
explode structs that are unrelated to tensorflow value types. One such example
is $Array<Int32>, which is a struct field of $TensorShape.

2. Extended SROA to allow skipping the explosion of alloc_stack over the
$TensorShape type.

While we may need to further enhance deabstraction code to propagate through
more SIL patterns (especially related to const array handling), the code is
structured in a way that's reasonably easy to extend.

mhong commented:
@swift-ci please test tensorflow

mhong commented on lib/SILOptimizer/Mandatory/TFDeabstraction.cpp:
```diff
@@ -1417,31 +1404,52 @@ propagateTensorOperand(SILValue v,
+  // %180 = apply %179<TensorShape>(%178) ...
+  // %181 = tuple_extract %180 : $(Array<TensorShape>, Builtin.RawPointer), 0
+  // %183 = tuple_extract %180 : $(Array<TensorShape>, Builtin.RawPointer), 1
+  // %185 = pointer_to_address %183
```
The array gets initialized by storing from %190 to %185. %181 is on the reader side of the array.

mhong commented on lib/SILOptimizer/Mandatory/TFDeabstraction.cpp:
```diff
@@ -1417,31 +1404,52 @@ propagateTensorOperand(SILValue v,
+  // Note that it is not sufficient to just process the operands of tensor ops.
+  //
+  // Example: need to propagate %188 to %190 (used as a tfop attr), where %189
+  // is not a const struct. This is so that const expr eval will only process
```
"This is so" should be correct. The work been done here will help the subsequent call "constantEvaluator.computeConstantValues(valuesToCheck, results)".

Reworded the comment here a bit.

Commit:
Polished comments.

mhong commented:
CI passed: https://ci-external.swift.org/view/Pull%20Request/job/swift-PR-TensorFlow-Linux/661/

mhong commented:
@lattner Adjusted this patch based on our discussion. PTAL.

lattner reviewed (approved):
I didn't have time for a detailed review, but a quick look looks good.  Thank you!

mhong commented:
Thanks for the review!
